lsp-bridge: init at 20230104

fmarl opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Lsp-bridge's goal is to become the fastest LSP client in Emacs.
Lsp-bridge uses python's threading technology to build caches that bridge Emacs and LSP server. Lsp-bridge will provide smooth completion experience without compromise to slow down emacs' performance.
So it is an alternative to the well known lsp-mode. Unfortunately there is no melpa/elpa package yet and to manage it in a reasonable way under NixOS I decided to write a package for it. The source code is available on [github](https://github.com/manateelazycat/lsp-bridge).
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

I am new and this is my first package. I hope I have fulfilled all requirements so far and I would have liked to enter myself as maintainer under meta, unfortunately nixpkgs-review threw an error. How do I become a maintainer?

Commit:
Adding nixpkg lsp-bridge to emacsPackages

Commit:
Fixing install phase of lsp-bridge

Commit:
Merge branch 'NixOS:master' into master

Commit:
Patching source and add python packages to lsp_bridge.py
The patch switches the default lsp-bridge-python-command to bash to run the wrapper

Commit:
Removing maintainer list from package

Commit:
Updating meta description in lsp-bridge

Commit:
Removing trailing whitespaces in lsp-bridge/default.nix

adisbladis commented:
The code looks fine but could you squash your commit history according to the contribution guidelines?

adisbladis commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/lsp-bridge/default.nix:
```diff
@@ -0,0 +1,78 @@
+{ lib
+, trivialBuild
+, fetchFromGitHub
+, python310Packages
```
Please use `python3` without minor qualifiers if that works.

fxttr replied:
Thanks for the hint, done that.

adisbladis commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/lsp-bridge/default.nix:
```diff
@@ -0,0 +1,78 @@
+    sha256 = "sha256-sB5niigN0rdtqeprlZAJEKgAuQDkcUMbbL9yTnrdoLg=";
+  };
+
+  patches = [ ./wrapper.patch ];
```
Could this be replaced with a `substituteInPlace` call instead of a patch?
Patches tend to break more often than a find/replace.

adisbladis commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/lsp-bridge/default.nix:
```diff
@@ -0,0 +1,78 @@
+
+  postInstall = with python310Packages; ''
+          wrapProgram $out/share/emacs/site-lisp/lsp_bridge.py \
+              --prefix PYTHONPATH : "${python310}" \
+              --prefix PYTHONPATH : "${epc}" \
+              --prefix PYTHONPATH : "${orjson}" \
+              --prefix PYTHONPATH : "${sexpdata}" \
+              --prefix PYTHONPATH : "${six}";
```
This is not going to fly.

`PYTHONPATH` leaks dependencies via env vars which also leaks into other processes launched by emacs.

It's better to use a `withPackages` closure as that is more hermetic.

fxttr replied:
Understood, didn't knew that. My first attempt to solve this, before using this wrapper, was to use propagatedBuildInputs. But it doesn't pull in these python packages as runtime dependeny, even when using

`
(python3.withPackages (ps: with ps; [ epc orjson sexpdata six ]))
`

Do you have a tip for me or could me point in the right direction?

adisbladis replied:
> Understood, didn't knew that.

Very few people do, it's one of really sharp edges integrating Nix/Python. 

This is totally untested, but something along these lines:
``` nix
let
  python = python3.withPackages (ps: with ps; [ epc orjson sexpdata six ]);
in
stdenv.mkDerivation {
  postPatch = ''
    substituteInPlace lsp-bridge.el --replace '(defcustom lsp-bridge-python-command (if (memq system-type '(cygwin windows-nt ms-dos)) "python.exe" "python3")' '(defcustom lsp-bridge-python-command "${python.interpreter}")'
  '';
}
```

That way lsp-bridge simply executes the wrapped interpreter and then we don't have to deal with `PYTHONPATH` at all :)

fxttr replied:
This works like a charm. Thanks for your great support! :)

Commit:
lsp-bridge: init at 20230104
- Switching from wrapper to python3.withPackages
- Replace the patch with substituteInPlace
- Generalizing python3 version

Baitinq commented:
@fxttr I believe squashing is still needed as per the guidelines :)

SuperSandro2000 commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/lsp-bridge/default.nix:
```diff
@@ -0,0 +1,66 @@
+
+  meta = {
+    description = "Fastest LSP client in Emacs.";
+    longDescription = ''
+        Using python's threading technology to build caches that bridge Emacs and LSP server.
+      '';
```
```suggestion
    longDescription = ''
      Using python's threading technology to build caches that bridge Emacs and LSP server.
    '';
```
missing homepage and maintainer

SuperSandro2000 commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/lsp-bridge/default.nix:
```diff
@@ -0,0 +1,66 @@
+  '';
+
+  postPatch = ''
+    substituteInPlace lsp-bridge.el --replace '(defcustom lsp-bridge-python-command (if (memq system-type '"'"'(cygwin windows-nt ms-dos)) "python.exe" "python3")' '(defcustom lsp-bridge-python-command "${python.interpreter}"'
```
```suggestion
    substituteInPlace lsp-bridge.el \
      --replace '(defcustom lsp-bridge-python-command (if (memq system-type '"'"'(cygwin windows-nt ms-dos)) "python.exe" "python3")' '(defcustom lsp-bridge-python-command "${python.interpreter}"'
```

SuperSandro2000 commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/lsp-bridge/default.nix:
```diff
@@ -0,0 +1,66 @@
+    sha256 = "sha256-sB5niigN0rdtqeprlZAJEKgAuQDkcUMbbL9yTnrdoLg=";
+  };
+
+  packageRequires =
+    [
+      posframe
+      markdown-mode
+      yasnippet
+      org
+      which-key
+    ];
```
```suggestion
  packageRequires = [
    posframe
    markdown-mode
    yasnippet
    org
    which-key
  ];
```

SuperSandro2000 commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/lsp-bridge/default.nix:
```diff
@@ -0,0 +1,66 @@
+  pname = "lsp-bridge";
+  version = "20230104";
+
+  commit = rev;
```
Is that used by emacs? Does not look like to me https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/emacs/trivial.nix

SuperSandro2000 commented on pkgs/applications/editors/emacs/elisp-packages/manual-packages/lsp-bridge/default.nix:
```diff
@@ -0,0 +1,66 @@
+  python = python3.withPackages (ps: with ps; [ epc orjson sexpdata six ]);
+in trivialBuild {
+  pname = "lsp-bridge";
+  version = "20230104";
```
```suggestion
  version = "unstable-2023-01-04";
```

kira-bruneau replied:
If using `trivialBuild`, `unstable-YYYY-MM-DD` is probably the best choice to fit in with the rest of nixpkgs, but if we want to generate ELPAs we need to fit the format expected by Emacs: https://github.com/emacs-mirror/emacs/blob/master/lisp/subr.el#L6670-L6751 (which is typically just the date shoehorned to fit into a MAJOR.MINOR version)

jian-lin commented:
Thanks for your work.

I just find this old attempt for lsp-bridge after reviewing a new one #219993.  IMHO, #219993 is in a better shape.  @kira-bruneau Maybe you can take some advice here such as https://github.com/NixOS/nixpkgs/pull/209652#discussion_r1064124812.

kira-bruneau commented:
Oh no, I didn't realize there was a PR open already  :disappointed:. I think I tried searching for "lspbridge" instead of "lsp-bridge" (which gave no results) and I had to [submit a patch upstream](https://github.com/manateelazycat/lsp-bridge/pull/557) before I could get byte compilation working, so I thought no one tried this yet :confused: 

@jian-lin Yep I could match how patching is done here in my PR. I was actually thinking about add a general `patchDefaults` option that would automatically update the defaults of any `defvar` - but maybe for now it would be easier to do a simple find and replace without having to fully parse elisp.

fxttr commented:
Funny coincidence, I sat down 2 hours ago to get the PR running here and squash my commits. I also ran into the byte compilation issues, as you can see I solved them less cleanly. Your approach is to be preferred in any case!

kira-bruneau commented:
> I am new and this is my first package. I hope I have fulfilled all requirements so far and I would have liked to enter myself as maintainer under meta, unfortunately nixpkgs-review threw an error. How do I become a maintainer?

For my PR I can also add you to the [maintainers list](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix) and add you as a maintainer if you'd like!

I've noticed in myself that I'm not as good as keeping packages up-to-date as I am at initially packaging them :sweat_smile: - and it would be nice to have more help while I try to figure out a better way to automate things.

fxttr commented:
With pleasure! :) I am glad if I can help.

kira-bruneau commented:
Replaced with #219993
